file/sequence_file_reader.cc

```diff
@@ -48,53 +62,87 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+      }
+      io_s =
+          file_->PositionedRead(aligned_offset + buf.CurrentSize(), allowed,
+                                IOOptions(), &tmp, buf.Destination(), nullptr);
+      if (ShouldNotifyListeners()) {
+        auto finish_ts = FileOperationInfo::FinishNow();
+        NotifyOnFileReadFinish(orig_offset, tmp.size(), start_ts, finish_ts,
+                               io_s);
+      }
+      buf.Size(buf.CurrentSize() + tmp.size());
+      if (!io_s.ok() || tmp.size() < allowed) {
+        break;
+      }
     }
-    io_s = file_->PositionedRead(aligned_offset, size, IOOptions(), &tmp,
-                                 buf.BufferStart(), nullptr);
-    if (io_s.ok() && offset_advance < tmp.size()) {
-      buf.Size(tmp.size());
+
+    if (io_s.ok() && offset_advance < buf.CurrentSize()) {
       r = buf.Read(scratch, offset_advance,
-                   std::min(tmp.size() - offset_advance, n));
+                   std::min(buf.CurrentSize() - offset_advance, n));
     }
     *result = Slice(scratch, r);
-    if (ShouldNotifyListeners()) {
-      auto finish_ts = FileOperationInfo::FinishNow();
-      NotifyOnFileReadFinish(orig_offset, tmp.size(), start_ts, finish_ts,
-                             io_s);
-    }
 #endif  // !ROCKSDB_LITE
   } else {
     // To be paranoid, modify scratch a little bit, so in case underlying
-    // FileSystem doesn't fill the buffer but return succee and `scratch`
+    // FileSystem doesn't fill the buffer but return success and `scratch`
     // returns contains a previous block, returned value will not pass
     // checksum.
     // It's hard to find useful byte for direct I/O case, so we skip it.
     if (n > 0 && scratch != nullptr) {
       scratch[0]++;
     }
 
+    size_t read = 0;
+    while (read < n) {
+      size_t allowed;
+      if (rate_limiter_priority != Env::IO_TOTAL && rate_limiter_ != nullptr) {
+        allowed = rate_limiter_->RequestToken(n - read, 0 /* alignment */,
+                                              rate_limiter_priority, nullptr,
+                                              RateLimiter::OpType::kRead);
+      } else {
+        allowed = n;
+      }
 #ifndef ROCKSDB_LITE
-    FileOperationInfo::StartTimePoint start_ts;
-    if (ShouldNotifyListeners()) {
-      start_ts = FileOperationInfo::StartNow();
-    }
+      FileOperationInfo::StartTimePoint start_ts;
+      if (ShouldNotifyListeners()) {
+        start_ts = FileOperationInfo::StartNow();
+      }
 #endif
-
-    io_s = file_->Read(n, IOOptions(), result, scratch, nullptr);
-
+      Slice tmp;
+      io_s = file_->Read(allowed, IOOptions(), &tmp, scratch + read, nullptr);
 #ifndef ROCKSDB_LITE
-    if (ShouldNotifyListeners()) {
-      auto finish_ts = FileOperationInfo::FinishNow();
-      size_t offset = offset_.fetch_add(result->size());
-      NotifyOnFileReadFinish(offset, result->size(), start_ts, finish_ts, io_s);
-    }
+      size_t offset = offset_.fetch_add(tmp.size());
```

Comment:
This line should go inside of `if (ShouldNotifyListeners()){}` since no reference to it outside of this if block

Reply:
Updated.